src/target/source/codegen_params.cc

```diff
@@ -0,0 +1,297 @@
+        for (int i = 0; i < num_elements; i++) {
+          int64_t elem = static_cast<int64_t*>(tensor->dl_tensor.data)[i];
+          uint64_t to_print;
+          if (elem < 0) {
+            os << "-";
+            to_print = -elem;
+          } else {
+            os << "+";
+            to_print = elem;
+          }
+          os << "0x" << std::setw(16) << to_print;
+          if (i < num_elements - 1) {
+            os << ", ";
+          }
+          if (((i + 1) % elements_per_row) == 0) {
+            os << "\n" << indent_str;
+          }
+        }
+      } else {
+        CHECK(false) << "should not get here";
+      }
+      break;
+
+    case runtime::DataType::TypeCode::kUInt:
+      CHECK(arr_type.bits() == 8 || arr_type.bits() == 16 || arr_type.bits() == 32 ||
+            arr_type.bits() == 64)
+          << "CodegenParams: only support generating 8-, 16-, 32-, or 64-bit integer params; saw "
+          << arr_type.bits() << "-bit array";
+
+      if (arr_type.bits() == 8) {
+        for (int i = 0; i < num_elements; i++) {
+          // NOTE: for special types int8_t and uint8_t, need to promote to int type to avoid
+          // printing as a char.
+          os << "0x" << std::setw(2)
+             << +static_cast<std::uint8_t>(static_cast<uint8_t*>(tensor->dl_tensor.data)[i]);
+          if (i < num_elements - 1) {
+            os << ", ";
+          }
+          if (((i + 1) % elements_per_row) == 0) {
+            os << "\n" << indent_str;
+          }
+        }
+      } else if (arr_type.bits() == 16) {
+        for (int i = 0; i < num_elements; i++) {
+          os << "0x" << std::setw(4) << static_cast<uint16_t*>(tensor->dl_tensor.data)[i];
+          if (i < num_elements - 1) {
+            os << ", ";
+          }
+          if (((i + 1) % elements_per_row) == 0) {
+            os << "\n" << indent_str;
+          }
+        }
+      } else if (arr_type.bits() == 32) {
+        for (int i = 0; i < num_elements; i++) {
+          os << "0x" << std::setw(8) << static_cast<uint32_t*>(tensor->dl_tensor.data)[i];
+          if (i < num_elements - 1) {
+            os << ", ";
+          }
+          if (((i + 1) % elements_per_row) == 0) {
+            os << "\n" << indent_str;
+          }
+        }
+      } else if (arr_type.bits() == 64) {
+        for (int i = 0; i < num_elements; i++) {
+          os << "0x" << std::setw(16) << static_cast<uint64_t*>(tensor->dl_tensor.data)[i] << "UL";
+          if (i < num_elements - 1) {
+            os << ", ";
+          }
+          if (((i + 1) % elements_per_row) == 0) {
+            os << "\n" << indent_str;
+          }
+        }
+      } else {
+        CHECK(false) << "should not get here";
+      }
+      break;
+
+    case runtime::DataType::TypeCode::kFloat: {
+      std::stringstream ss;
+      ss.setf(std::ios::hex | std::ios::showbase | std::ios::fixed | std::ios::scientific,
+              std::ios::basefield | std::ios::showbase | std::ios::floatfield);
```

Comment:
`std::showbase` is there twice.

Reply:
right, the second one is a mask so I think it's needed twice?